klippy/extras/safe_power.py

```diff
@@ -0,0 +1,116 @@
+        for s in config.get_prefix_sections('manual_stepper'):
+            self._register_stepper(s, ppins)
+        if config.has_section('dual_carriage'):
+            self._register_stepper(config.getsection('dual_carriage'), ppins)
+        #Check config order
+        for name, obj in self.printer.lookup_objects():
+            if 'tmc' in name.lower():
+               raise self.printer.config_error(
+                     "[%s] section should be before"
+                     " any tmc section in printer.cfg" % (self.name,))
+        self.mcu.register_config_callback(self._build_config)
+        self.printer.register_event_handler("klippy:connect",
+                                            self._connect)
+    def _build_config(self):
+        self.set_digital_out_cmd = self.mcu.lookup_command(
+            "set_digital_out pin=%u value=%c")
+        self.read_digital_in_cmd = self.mcu.lookup_query_command(
+            "read_digital_in pin=%u pull_up=%c",
+            "read_digital_in_value pin=%u value=%c")
+    def _connect(self):
+        ucmd = self.set_digital_out_cmd.send
+        res = None
+        # Disabling mcu power
+        ucmd([self.pin, 0])
+        if len(self.steppers) > 0:
+            logging.info("%s starting checks for stepper"
+                         " driver backwards" % (self.name,))
+            for stepper in self.steppers.values():
+                res = self._check_driver(stepper)
+                if res is not None:
+                    break
+        if res is not None:
+            raise self.error(res)
+        else:
+            # Enabling mcu power
+            ucmd([self.pin, 1])
+    def _register_stepper(self, config, ppins):
+        name = config.get_name()
+        enable_pin = config.get('enable_pin', None)
+        step_pin = config.get('step_pin', None)
+        # Only register stepper driver with enable and step pins
+        if enable_pin is not None and step_pin is not None:
+            enable_pin_params = ppins.parse_pin(enable_pin, can_invert=True)
+            step_pin_params = ppins.parse_pin(step_pin, can_invert=True)
+            # Only register stepper driver belongs to current mcu
+            if (step_pin_params['chip'] is self.mcu
+                    and enable_pin_params['chip'] is self.mcu):
+                self.steppers[name] = StepperTracking(
+                    name, step_pin_params, enable_pin_params)
+    def _check_driver(self, stepper):
+        logging.info("Checking [%s]" % (stepper.name,))
+        res = None
+        # Reconfig stepper driver pins
+        curtime = self.reactor.monotonic()
+        print_time = self.mcu.estimated_print_time(curtime)
+        minclock = self.mcu.print_time_to_clock(print_time + .100)
+        self.set_digital_out_cmd.send([stepper.step_pin_params['pin'],0],
+                                      minclock=minclock)
+        # Do the stepper driver check
+        minclock = self.mcu.print_time_to_clock(print_time + .100)
```

Comment:
This `minclock` line will return an identical value as the previous `minclock` calculation.  Maybe you want `print_time + 0.200` to ensure a 100ms delay since the pin was set as output?